lms/djangoapps/courseware/docs/decisions/0004-personalized-relative-dates.rst

```diff
@@ -0,0 +1,71 @@
+self-paced courses are currently only able to offer the latter experience.
+
+``edx-when`` is a library currently used by ``edx-platform`` to provide a
+read-optimized view of the dates associated with course content. It is updated
+when a course is published, and stores dates in a dedicated table in the
+relational database.
+
+~~~~~~~~
+Decision
+~~~~~~~~
+
+In order to support incremental deadlines in self-paced courses, we want
+to extend edx-platform with the following capabilities:
+
+1.  Extend the usage of edx-when in the LMS to begin reading the ``rel_date``
+    field, and convert it when loading by interpreting it as an offset
+    from the users ``start`` as defined by their ``Schedule``.
+2.  When a self-paced course is published, add relative due dates to it,
+    evenly spaced across the expected duration of the course, unless the
+    course already has relative dates attached to content.
+
+    a.  Question: should these dates should be pushed into ``edx-when``
+        by Studio, rather than being part of the ``edx-when`` course publish
+        signal processing, because it is a specific business case? Or should
+        ``edx-when`` include that level of business knowledge because it
+        generally gives better results?
+    b.  While there are no current plans to allow Studio authors to author
+        relative dates, by making the evenly-spacing only occur if there
+        is no existing relative dates on the course, we leave ourselves the
+        ability to add future studio authoring without changing ``edx-when``.
+    c.  Adding the relative dates on course-publish, and storing them in ``edx-when``,
+        rather than inferring them at run-time while reading from ``edx-when``
+        means that we have the course structure in-hand (from the course
+        publish), and that ``edx-when`` will remain a full listing of all dates
+        in the course. If, instead, the LMS inferred relative dates when they
+        weren't set in ``edx-when``, then new date-based functionality in the
+        future would need to either read through a specific api that would
+        inject the programmatic dates, or would have to infer its own dates.
+3.  No new fields would be added to XBlocks at this time to represent relative
+    dates. There is no studio authoring component for this feature, and the LMS
+    Xblock runtime would contextualize relative dates for a specific student
+    when supplying them to XBlock fields.
+
+~~~~~~
+Status
+~~~~~~
+
+Proposed
+
+~~~~~~~~~~~~
+Consequences
+~~~~~~~~~~~~
+
+Once implemented, self-paced courses in the LMS would begin to have
+due dates for course content that are consistent between users, but that
+are personalized based on when a user started a course. We would be open
+to exploring future functionality for allowing users to adjust their schedule
+by resetting their effective course start time, or speeding up or slowing
+down their pacing.
```

Comment:
I think it's worth considering these future capabilities in today's discussion, to set up the extensibility we're going to need a fast-follow to the initial functionality